### PR TITLE
Code for generating error catalog from source code

### PR DESCRIPTION
**Pull request by davidcok**
<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/Compilers/CompilerErrors.cs [hunk 1]
```diff
@@ -53,150 +53,333 @@ public enum ErrorId {
...
+such as the proper files not having the correct permissions.
 ");
 
     Add(ErrorId.c_unsupported_feature,
     @"
+This messages indicates two problems:
+- the given feature is not supported in the compiler for the target language but is present in the program,
+so the program will need to be revised to avoid this feature;
+- the feature is not listed in the in-tool list of unsupported features.
+The latter is a (minor) bug in the in-tool documentation. Please report this error message and the part of the
```
Comment (atomb) [inline Source/DafnyCore/Compilers/CompilerErrors.cs:86]:
```suggestion
The latter is an omission in the in-tool documentation. Please report this error message and the part of the
```

File: Source/DafnyCore/Compilers/CompilerErrors.cs [hunk 2]
```diff
@@ -53,150 +53,333 @@ public enum ErrorId {
...
+program provoking it to the Dafny team's [issue tracker](https://github.com/davidcok/dafny/issues).
```
Comment (atomb) [inline Source/DafnyCore/Compilers/CompilerErrors.cs:87]:
```suggestion
program provoking it to the Dafny team's [issue tracker](https://github.com/dafny-lang/dafny/issues).
```

File: Source/DafnyCore/Compilers/CompilerErrors.cs [hunk 3]
```diff
@@ -53,150 +53,333 @@ public enum ErrorId {
...
+
+If the assumption marked with `{:axiom}` is not actually valid, then the validity of the entire program is in question.
 ");
 
     Add(ErrorId.c_forall_statement_has_no_body,
     @"
+A method may be parsed and verified even if a [forall statement](../DafnyRef/DafnyRef#sec-forall-statement) is missing a body. 
+However, the body must be supplied before the program can be compiled,
+even if the method is `ghost`. Body-less foralls in ghost methods are 
+similar to unchecked assumptions.
```
Comment (atomb) [inline Source/DafnyCore/Compilers/CompilerErrors.cs:244]:
```suggestion
unchecked assumptions.
```

Is there any way in which they aren't exactly that?

File: Source/DafnyCore/Compilers/CompilerErrors.cs [hunk 4]
```diff
@@ -53,150 +53,333 @@ public enum ErrorId {
...
 ");
 
     Add(ErrorId.c_loop_has_no_body,
     @"
+A method may be parsed and verified even if a loop is missing a body. 
+Note that a missing body is different than an empty body, which is just `{ }`.
+However, the body must be supplied before the program can be compiled,
+even if the method is `ghost`. 
+Body-less loops in ghost methods are similar to unchecked assumptions.
```
Comment (atomb) [inline Source/DafnyCore/Compilers/CompilerErrors.cs:253]: Here, "similar" may be more accurate.

File: Source/DafnyCore/Compilers/CompilerErrors.cs [hunk 5]
```diff
@@ -53,150 +53,333 @@ public enum ErrorId {
...
     Add(ErrorId.c_binding_if_forbidden,
     @"
+The `--enforce-determinism` option requires a target program to be deterministic and predictable:
+there may be no program statements that have an arbitrary, even if deterministic, result.
+The [`if` with a binding guard](../DafnyRef/DafnyRef#sec-binding-guards) (`if ... :| `) checks that the 
+given predicate is satisfiable by some value. If not, then the 'else' branch is executed;
+but if so, the 'then' branch is executed with an arbitrary value that satisifies the predicate.
+Because of this arbitrary selection, the if-with-binding-guard is not permitted with `--enforce-determinism`,
+even if there is exactly one value that satisfies the predicate.
+(The tool does not try to determine whether there is just one value and
```
Comment (atomb) [inline Source/DafnyCore/Compilers/CompilerErrors.cs:300]:
```suggestion
(The tool does not try to determine whether there is just one value or
```

File: Source/DafnyCore/Compilers/CompilerErrors.cs [hunk 6]
```diff
@@ -53,150 +53,333 @@ public enum ErrorId {
...
+whether there is a reasonable way to compute it.)
```
Comment (atomb) [inline Source/DafnyCore/Compilers/CompilerErrors.cs:301]:
```suggestion
whether there is a reasonable way to compute a value.)
```

**Review by atomb — CHANGES_REQUESTED** [review]
I saw a few tiny wording things that could possibly be improved, but this otherwise looks great.

**Review by atomb — DISMISSED** [review]
Looks like you still need to check in the generated error catalog, but otherwise 🚢 .

**Comment by davidcok** [comment]
Yep - first time using the process, so you need to push the approve button again. -- Optimistically presuming the tests will succeed

**Review by atomb — APPROVED** [review]

**Pull request merged**